### DISCUSSION TITLE
How to install hydrogym on a supercomputer ?

### DISCUSSION THREAD

**Discussion by ReHoss** [dynamicslab/hydrogym#67] [Q&A]
Hi everyone, since firedrake is not conda friendly, what are the best practices to install hydrogym on a supercomputer ? 
I mean there is this effort to include conda compatibility : https://github.com/firedrakeproject/firedrake/pull/2543#issue-1359598265

Otherwise, (I am new to HPC) could it be enough to work with a pip environment (the one created during firedrake installation) in the HPC node ?

Or should we use Docker ?

**Comment by jcallaham** [comment]
The best route will probably depend a little on how packages and permissions and such are managed on your cluster.  But I think that probably the more difficult thing to install will be Firedrake compared to Hydrogym.  Have you been able to get Firedrake working, or does there happen to be a module with it pre-installed on your cluster?

From there you can just install Hydrogym into whatever environment works for Firedrake.  For instance, if you are able to get Firedrake working with conda you can just clone Hydrogym locally and then do `python -m pip install -e .`, which should install into your conda environment using the conda Python version.

Docker is also a good alternative, if that's something that works well on your cluster.  Just depends on what the admins prefer.

**Reply by ReHoss** [reply]
Hello @jcallaham , I am just anticipating. I agree with you on Firedrake would probably be the most difficult step.

I have access to different clusters, each with their own permissions I have not figured out yet... 

But indeed the best way would be to replicate the installation I have done on my personal laptop, by pulling Firedrake, use the installation script and use the venv created during installation.

I'll post an update once I would have give some tries.


In your case, does Firedrake is available on your cluster or do you install it like on your laptop ? I am afraid the installation of Firedrake requires permission to install packages directly onto the cluster system ?


Thank you for your availability:)

**Reply by jcallaham** [reply]
That would be great if you update once you figure out some solutions - I think the permissions and such can vary quite a bit, so it might help other potential users.

For me, I haven't run it on a cluster yet.  We are working on getting access for collaborators, but I've just been doing development on a workstation that has plenty of compute for normal simulations (i.e. not RL training).

**Reply by ReHoss** [reply]
The potential solutions:

- Ask to the HPC administrators to install Firedrake (the one I am following since second solution fails for now)


- Firedrake devs developed a Spack environment which does not requires super-user rights https://github.com/firedrakeproject/firedrake-spack .
However I tried this and it failed because of a relatively minor bug on PETSc MPI linkage https://github.com/firedrakeproject/firedrake-spack/issues/28 . Moreover, I don't really know how to bind hydrogym to this Spack environments and other python libraries I would like to use...


- Use a Docker/Singularity container
My HPC use Singularity which handle Docker containers. However I don't know how easy it is to run _Job _Arrays__ when having the program within a container, I mean do I need to duplicate the Docker image n times for n differents experiments ?
plus I don't know if the container is able to write to my $WORKDIR database directly (I guess it is possible with port etc.?).

Thanks:)

**Reply by ludgerpaehler** [reply]
Hey,

In the update we are currently pushing up we went with Option 3 where Ray launches container instances, and Ray itself can be run in the cloud on a Kubernetes-based system, or talk to a Slurm-based cluster. You should only have to wait for ~1 week until the final state will be verified, and pushed up into the repo.

Best,
Ludger

**Reply by ReHoss** [reply]
Hi @ludgerpaehler @jcallaham , I had to change of HPC and know it becomes difficult to me to install FireDrake again on a new HPC, is there a Singularity solution or Ray available now?

**Reply by ludgerpaehler** [reply]
Can you remind me at the end of the week :) Currently working to update the docs, which should then hopefully be able to answer all your questions and if not merit more work from my side :)
